core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala

```diff
@@ -75,21 +95,31 @@ private[memory] class StorageMemoryPool(lock: Object) extends MemoryPool(lock) w
   def acquireMemory(
       blockId: BlockId,
       numBytesToAcquire: Long,
-      numBytesToFree: Long): Boolean = lock.synchronized {
+      numBytesToFree: Long): Boolean = {
     assert(numBytesToAcquire >= 0)
     assert(numBytesToFree >= 0)
-    assert(memoryUsed <= poolSize)
-    if (numBytesToFree > 0) {
+    assert(lock.synchronized(memoryUsed <= poolSize))
+    val memoryFreedByEviction = if (numBytesToFree > 0) {
       memoryStore.evictBlocksToFreeSpace(Some(blockId), numBytesToFree)
+    } else {
+      0
     }
-    // NOTE: If the memory store evicts blocks, then those evictions will synchronously call
-    // back into this StorageMemoryPool in order to free memory. Therefore, these variables
-    // should have been updated.
-    val enoughMemory = numBytesToAcquire <= memoryFree
-    if (enoughMemory) {
-      _memoryUsed += numBytesToAcquire
+    if (memoryFreedByEviction > numBytesToAcquire) {
+      releaseMemory(memoryFreedByEviction - numBytesToAcquire)
+    }
```

Comment:
+1, also I'll just wrap both things in synchronized so we don't have to acquire the lock twice. Fewer interleavings to think about.